src/libxtp/gwbse/bse.cc

```diff
@@ -655,5 +660,93 @@ BSE::Interaction BSE::Analyze_eh_interaction(const QMStateType& type,
   return analysis;
 }
 
+void BSE::Perturbative_DynamicalScreening(
+    const QMStateType& type, const Orbitals& orb,
+    const Eigen::VectorXd& Hd_static_contrib) {
+
+  Interaction analysis;
+
+  const tools::EigenSystem& BSECoefs =
+      (type == QMStateType::Singlet) ? orb.BSESinglets() : orb.BSETriplets();
+
+  const Eigen::VectorXd& RPAInputEnergies = orb.RPAInputEnergies();
+  const Eigen::VectorXd& BSEenergies = BSECoefs.eigenvalues();
+
+  // initial copy of static BSE energies to dynamic
+  Eigen::VectorXd BSEenergies_dynamic = BSEenergies;
+
+  // recalculate Hd at the various energies
+  for (Index i_exc = 0; i_exc < BSEenergies.size(); i_exc++) {
+    XTP_LOG(Log::info, _log) << "Dynamical Screening BSE, Excitation " << i_exc
+                             << " static " << BSEenergies(i_exc) << flush;
+
+    for (Index iter = 0; iter < _max_dyn_iter; iter++) {
+
+      // setup the direct operator with the last energy as screening frequency
+      double old_energy = BSEenergies_dynamic(i_exc);
+      SetupDirectInteractionOperator(RPAInputEnergies, old_energy);
+      HdOperator hd_dyn(_epsilon_0_inv, _Mmn, _Hqp);
+      configureBSEOperator(hd_dyn);
+
+      // get the contribution of Hd for the dynamic case
+      Interaction analysis_dyn;
+      /* this is a bit superfluous because we are only interested in the
+         contribution of Hd_dyn to a specific excitation, not all of them, but
+         leave for later */
+      analysis_dyn.direct_contrib =
+          Analyze_IndividualContribution(type, orb, hd_dyn);
+
+      // new energy perturbatively
+      BSEenergies_dynamic(i_exc) = BSEenergies(i_exc) +
+                                   Hd_static_contrib(i_exc) -
+                                   analysis_dyn.direct_contrib(i_exc);
```

Comment:
This part is confusing, why should the analysis be used for perturbation theory? Why not use a a second HdOperator ?

Reply:
I see you want to save the calculation for the static case. Considering that the saving from that will be pretty small. I would recalculate the static case here at the beginning. It does not cost much and it separates the code nicely.

Reply:
It will cost you one iteration.

Reply:
The analysis function is used because what we need to calculate is the expectation value of Hd(dynamic) with the static BSEcoefs. And this function provides it.

Reply:
However, it calculates this for all S, not only the one we are interested in, which is a bit useless.

Reply:
Yes, calculating the expectation value should be simply `A.T*H*A`
so I think not using the function is cleaner.

Reply:
it might be that `H` does not provide a simple overload for matrix*vector operations only matrix*matrix

Reply:
Factoring out a function to calculate an expectation value is probably better then using the analysis function, because the name `analysis` would be misleading.

Reply:
But `Analyze_IndividualContribution` is exactly such a function that calculates the expectation value(s). Maybe we could just rename it to avoid confusion?

Reply:
yes, that is a good idea, but we should change the function signature then as well. Using an eigen::system and the operator should be enough. I think you also do not need the interaction struct.

Reply:
OK, so I have renamed the variables inside the `Perturbative_DynamicalScreening` function. The static case can actually reuse `Analyze_IndividualContribution` because that gives us all `Hd` contributions for every S at the same time, which is exactly what we need (unless we want to add an option to just do this for a selected range of states...).

Not sure what you mean by

> we should change the function signature then as well. Using an eigen::system and the operator should be enough.

Reply:
> and of course the BTDA case has to be considered as well.

Indeed, and I think what `Analyze_eh_interaction` does now is maybe not correct for the full BSE, for example:

`S =    5 Omega = +13.661449286537 eV  lamdba = +90.77 nm <FT> = +22.0218 <K_x> = +0.7932 <K_d> = -9.8635`

but

`<FT> + 2<K_x> + <K_d> = 13.7447 eV` ?